138 - implement first and last. https://github.com/RoaringBitmap/Roar…

openkappa opened the pull request:
…ingBitmap/issues/138

Following on from comments on issue page this implementation adds first() and last() to container implementations. The offset issue is resolved within RoaringArray, which is delegated to by RoaringBitmap. Tests added for a range of values for each container implementation and RoaringBitmap itself.

lemire commented:
+1

gssiyankai commented:
LGTM

lemire commented:
@gssiyankai  Thanks. 

This needs to be ported to the buffer package before we can produce a release. Not hard. Mostly copy and paste. I can do it later... or maybe someone wants to jump in and do it... ;-) (It is probably a 5-minute job.)

openkappa commented:
I can do it later. I don't actually use the buffer package - is it effectively a carbon copy of the non persistent implementation?

lemire commented:
@openkappa 

Yes, it is a carbon copy that works over ByteBuffers instead of arrays. It is annoying that we need to maintain two distinct versions of the same code, but there are significant performance trade-offs involved. The ByteBuffer approach works with off-heap processing... so it can avoid the garbage collection... but it is slower compared to working directly on Java arrays.

Mostly, it is a minor annoyance involving copying and pasting.

openkappa commented:
Buffer version is much more verbose. The branch coverage is going to decrease with a port unless cases can be contrived where the buffer hasn't got an array. Any thoughts on contriving cases to cover array backed buffers and the converse?

Kineolyan commented on src/main/java/org/roaringbitmap/RoaringArray.java:
```diff
@@ -578,4 +578,30 @@ protected int size() {
+   */
+  public int first() {
+    if(size == 0) {
+      return 0;
```
Is this the good policy to return 0 as the value when there is no elements in the RoaringArray (or a RoaringBitmap) ?
I thought that 0 would be a valid value. We may simply throw an error `NoSuchElementException`, as Collections.max does in the JDK.

openkappa replied:
I agree. I'm not sure if Collections::max relates to this functionality, but SortedSet::first, SortedSet::last throw NoSuchElementException for empty sets. I must admit until now I had only considered bitmaps containing natural numbers, but to be consistent with the comments on the PositiveIterator thread the only option would be to throw.

Kineolyan replied:
Yes, Collections::max was the first that came to my mind.
I thought about 0 as a possible value as I know that when using bitmaps as index for lines of DB tables, those lines are refered by their index as in an array - starting from 0.
Anyway, I like it very much now :)

Kineolyan commented on src/test/java/org/roaringbitmap/TestRoaringBitmap.java:
```diff
@@ -4342,4 +4342,49 @@ public void testLazyOr() {
+      int x = 1 << i;
+      rb.add(x);
+      Assert.assertEquals(1, rb.first());
+      rb.last();
```
Is this call mandatory ? You are not storing the last value and computing it again on the next line.

Kineolyan commented:
Good job @openkappa

openkappa commented:
The last commit incurred a performance penalty for run containers. I've fixed that and added explicit tests for cases where the bitmap has been lazily mutated.

lemire commented:
LGTM.
